terrawrap/utils/graph.py

```diff
@@ -0,0 +1,108 @@
+""" Holds graph utilities"""
+
+import os
+from typing import List, Tuple, Any
+import networkx
+
+
+def has_cycle(graph: networkx.DiGraph) -> bool:
```

Comment:
This uses both networkx's built in cycle detection as well as checking if there is not source node in the directory. A source node is one that has no predecessors( no dependencies) and can be run by itself at anytime without breaking ex. IAM